airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQueryDestinationHandler.java

```diff
@@ -51,6 +56,37 @@ public boolean isFinalTableEmpty(final StreamId id) {
     return BigInteger.ZERO.equals(bq.getTable(TableId.of(id.finalNamespace(), id.finalName())).getNumRows());
   }
 
+  @Override
+  public Optional<Instant> getMinTimestampForSync(final StreamId id) throws Exception {
+    final Table rawTable = bq.getTable(TableId.of(id.rawNamespace(), id.rawName()));
+    if (rawTable == null) {
+      return Optional.empty();
+    }
+    final TableResult queryResult = bq.query(QueryJobConfiguration.newBuilder(new StringSubstitutor(Map.of(
+        "raw_table", id.rawTableId(BigQuerySqlGenerator.QUOTE))).replace(
+            // bigquery timestamps have microsecond precision
+            """
+            SELECT COALESCE(
+              (
+                SELECT TIMESTAMP_SUB(MIN(_airbyte_extracted_at), INTERVAL 1 MICROSECOND)
+                FROM ${raw_table}
+                WHERE _airbyte_loaded_at IS NULL
+              ),
+              (
+                SELECT MAX(_airbyte_extracted_at)
+                FROM ${raw_table}
+              )
+            )
+            """))
+        .build());
+    final FieldValue value = queryResult.iterateAll().iterator().next().get(0);
```

Comment:
Is the `iterator.next()` guaranteed to return single element, no threat of NPE ?

Reply:
yep! the query is just a simple `select <something>`, so it always returns exactly one record. The subqueries have the actual `select ... from ...` statements.